docs/custom_promise_types/modules.md

```diff
@@ -0,0 +1,367 @@
+
+A promise module will receive 3 types of requests (operations):
+
+1. `validate_promise` - Validate the names and types of promise attributes
+2. `evaluate_promise` - Perform changes to the system
+3. `terminate` - Shut down the module
+
+The module should not make any assumptions about the order of validation or evaluation.
+Each promise might be validated and evaluated, 0, 1, or more times, in any order.
+In general, the promise module should not require that state is maintained between the individual requests.
+
+The agent _might_ start multiple processes of the same module in parallel, for example for a module implementing multiple promise types.
+The module should not rely on this.
+Additionally, the same module may be started and stopped multiple times during an agent run, so don't make any assumptions that it will only run once per agent run.
+
+The promise module's responses to cf-agent requests must include `operation` and `result`.
+Promiser string and attributes are typically added for convenience.
+
+#### JSON protocol
+
+The JSON based protocol supports all aspects of CFEngine promises, including `slist`, `data` containers, and strings with newlines, tabs, etc.
+It is designed to be easy to implement in a language with JSON support.
+
+Protocol messages are separated by empty lines (double newline).
+The headers (request and response) are not JSON, but a sequence of space-separated values.
+All messages sent by cf-agent and the promise module are single line JSON-data, except:
+
+* Headers (both from cf-agent and promise module) are not JSON.
+* JSON responses sent from promise module may optionally be preceeded by log messages, as explained below.
+
+Within strings in the JSON data, newline characters must be escaped (`\n`).
+This is not strictly required by the JSON spec, but most implementations do this anyway.
+
+##### Example requests in JSON based protocol
+
+```
+CFEngine 3.16.0 v1
+
+{"operation": "validate_promise", "log_level": "info", "promise_type": "git", "promiser": "/opt/cfengine/masterfiles", "attributes": {"repo": "https://github.com/cfengine/masterfiles"}}
+
+{"operation": "evaluate_promise", "log_level": "info", "promise_type": "git", "promiser": "/opt/cfengine/masterfiles", "attributes": {"repo": "https://github.com/cfengine/masterfiles"}}
+
+{"operation": "terminate", "log_level": "info"}
+```
+
+##### Example response in JSON based protocol
+
+```
+git_promises 0.0.1 v1 json_based
+
+{"operation": "validate_promise", "promiser": "/opt/cfengine/masterfiles", "attributes": {"repo": "https://github.com/cfengine/masterfiles"}, "result": "valid"}
+
+log_info=Cloning 'https://github.com/cfengine/masterfiles' -> '/opt/cfengine/masterfiles'...
+log_info=Successfully cloned 'https://github.com/cfengine/masterfiles' -> '/opt/cfengine/masterfiles'
+{"operation": "evaluate_promise", "promiser": "/opt/cfengine/masterfiles", "attributes": {"repo": "https://github.com/cfengine/masterfiles"}, "result": "kept"}
```

Comment:
Is there particular reason for including both the promiser and the attributes? Do we expect the responses to arrive out of order?

If so, would it make sense to have some kind of `request_id`, which is then mirrored in the response? A promise can be listed in several different bundles, and it doesn't look like it will be possible to differentiate between responses if the promiser and attributes are the same.

Reply:
No, responses will not arrive out of order. I included them, because it made it easier to follow and debug, but we can remove them from the response.

Reply:
Ack